devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Add(*[-self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    def __isub__(self, other):
+        if isinstance(other, Differentiable):
+            self._expr = sympy.Add(*[self.expr, -other.expr], evaluate=False)
+        else:
+            self._expr = sympy.Add(*[self.expr, -other], evaluate=False)
+        return self
+
+    def __mul__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(sympy.Mul(*[self.expr, other.expr], evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Mul(*[self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    __rmul__ = __mul__
+
+    def __imul__(self, other):
+        if isinstance(other, Differentiable):
+            self._expr = sympy.Mul(*[self.expr, other.expr], evaluate=False)
+        else:
+            self._expr = sympy.Mul(*[self.expr, other], evaluate=False)
+        return self
+
+    def __truediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(self.expr * sympy.Pow(other.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(self.expr * sympy.Pow(other, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __rtruediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(other.expr * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(other * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __neg__(self):
+        return -1.0 * self
+
+    def __pow__(self, exponent):
+        return Differentiable(sympy.Pow(self, exponent, evaluate=False), fd=self.fd)
+
+    def __eq__(self, other):
+        if isinstance(other, Differentiable):
+            return self.expr.args == other.expr.args
+        elif isinstance(other, sympy.Expr):
+            return self.expr.args == other.args
+        else:
+            return sympy.simplify(self.expr) == other
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def func(self, *args, **kwargs):
+        kwargs.pop("evaluate", None)
+        if hasattr(self.expr, 'base'):
+            return self
+        else:
+            return Differentiable(self.expr.func(*args, **kwargs), fd=self.fd)
+
+    def __str__(self):
+        return self.expr.__str__()
+
+    __repr__ = __str__
+
+    @property
+    def args(self):
+        if hasattr(self.expr, 'base'):
+            return (self.expr,)
+        return self.expr.args
+
+    @property
+    def is_TimeFunction(self):
```

Comment:
shouldn't we call it ``is_TimeExpr`` rather than ``is_TimeFunction`` ?

Reply:
I think called it `is_TimeFunction` for dse reasons